### DISCUSSION TITLE
Docker DNS Service Discovery "no such host"

### DISCUSSION THREAD

**Discussion by SIMULATAN** [grafana/mimir#4517] [Help and support]
I'm running two mimir nodes in a stack and want to use DNS Service Discovery to connect the two nodes together.
Therefore, I configured `dns+tasks.monitoring_mimir` as a `join_member` in `mimir.yml`.
Unfortunately, when it tries to do the DNS query, this is printed in the console:
```
caller=resolver.go:87 level=error msg="failed to lookup IP addresses" host=tasks.monitoring_mimir err="lookup tasks.monitoring_mimir on 127.0.0.11:53: no such host"
caller=resolver.go:134 level=warn msg="IP address lookup yielded no results. No host found or no addresses found" host=tasks.monitoring_mimir
```

However, when running `dig tasks.monitoring_mimir`, it works just as expected!
![screenshot of the `dig` command output](https://user-images.githubusercontent.com/63104422/225601451-7f403bcd-fe6f-495f-b5bf-a679eb7279cb.png)
(both when doing a `docker exec` and using that as the entrypoint)

Also, when looking at the dashboard, I can see that the member for some reason uses the wrong interface?
That one is the bridge interface made when publishing ports.
![mimir web dashboard - memberlist cluster members](https://user-images.githubusercontent.com/63104422/225601749-02eede7f-b73f-4f0d-8773-38bb264e7ae6.png)
I tried setting the name manually to the correct interface name (`eth0`) by using `instance_interface_names`, but it's no use.

[Logfile](https://gist.github.com/SIMULATAN/4057c3d7002726608b3acf233b30abc7)
[mimir.yml](https://gist.github.com/SIMULATAN/54acf3681b4feae3a2522746298bba7c)
[docker-compose.yml](https://gist.github.com/SIMULATAN/a2674dddd370b7724d932df623b8426d)

I tried multiple combinations of using `instance_interface_names`, `instance_addr`, service discovery prefixes, network drivers, nodes, architectures, replication modes and more, but nothing worked, it keeps erroring with the above logs.

Another observation was that it tries to connect to some IPs in the `192.168.x.x` subnet, even though I didn't add that range anywhere and the DNS query doesn't return those IPs.

**Comment by pracucci** [comment]
> `caller=resolver.go:87 level=error msg="failed to lookup IP addresses" host=tasks.monitoring_mimir err="lookup tasks.monitoring_mimir on 127.0.0.11:53: no such host"`
> `caller=resolver.go:134 level=warn msg="IP address lookup yielded no results. No host found or no addresses found" host=tasks.monitoring_mimir`

I'm not sure what's happening here. The configuration you shared looks correct to me. Since I'm not familiar with your setup, let me ask a dumb question: can you confirm the host `tasks.monitoring_mimir` exists before Mimir process starts, right?

> Also, when looking at the dashboard, I can see that the member for some reason uses the wrong interface?
> That one is the bridge interface made when publishing ports.

The IP address exposed by memberlist can be configured setting `-memberlist.advertise-addr` or its respective YAML config option ([doc](https://grafana.com/docs/mimir/latest/references/configuration-parameters/#memberlist)).  The default is using the address from the same interface where the memberlist listening port was binded, which is configured by `-memberlist.bind-addr`.

**Reply by SIMULATAN** [reply]
Thanks for the response!
> can you confirm the host `tasks.monitoring_mimir` exists before Mimir process starts, right?

Well, depends. I did try setting the `dig` command as the entrypoint (=> it runs pretty much instantly, or, at least faster than mimir's DNS lookup) and it worked as expected.

About the advertise addr; can I also tell it to bind to a specific **interface**? Since I'm using docker and dynamically created networks I try to not hard code any IP (ranges) but rather let docker decide what IPs to give out. After restarting a few times, I did observe that the interface name stays consistent (even with multiple networks attached to the container). I saw that `instance_interface_names` exists for some objects but not for `KVConfig` apparently.
Also, I even if I hardcoded the range, I couldn't use these two settings as the cluster scales with docker swarm and thus the config is the same for all instances.

Do I even need advertising or can that just be bypassed by using DNS service discovery?

Oh, also, about the DNS lookup; as you can see in the screenshot, `tasks.monitoring_mimir` doesn't resolve to a specific single host but rather to multiple ones (including the current instance), I checked the docs on that but since `dig` seems to return `A` results and the query was an `A` one too `dns+` should be the right service discovery method, right..?

I can also observe that the second instance does pop up in the dashboard for around 30 seconds, after which it just disappears and is gone forever with the KV Store Values rising linearly.
![initial](https://user-images.githubusercontent.com/63104422/226701969-83991c0d-3118-4f53-b89d-b65a4176fdae.png)
![after 30 seconds](https://user-images.githubusercontent.com/63104422/226701986-a530f7f0-ef2e-4ec3-87e4-262fd2efe11d.png)


Anyway, I went ahead and put my logs into a gist to avoid overlooking important details.
https://gist.github.com/SIMULATAN/c529220bc236e6f8e04332777d215c3e

Thanks a ton for the help!

**Reply by SIMULATAN** [reply]
Well, I managed to fix the IP issue by introducing a hack that allows this whole thing to be dynamically scalable but still work just fine.

I just used this as an entrypoint:
`/entrypoint.sh`
```sh
#!/bin/sh
IP="$(ip addr show eth1 | grep inet | tr -s ' ' | cut -f 3 -d ' ' | cut -f 1 -d '/')"
cat /etc/mimir/config.yml | sed 's/%IP%/'"$IP"'/' | tee /etc/mimir/config-final.yml
/bin/mimir --config.file=/etc/mimir/config-final.yml --config.expand-env
```

This checks the interface of the overlay network, gets the IP from it and substitutes all cases of `%IP%` in my `mimir.yml`
This works because I set all the `instance_addr`s available to `%IP%` to be substituted with the actual IP when the container boots up.

Now, the instances connect together just fine!
![image](https://user-images.githubusercontent.com/63104422/231105825-f447b490-99de-41ea-937f-e39a0aad9825.png)
When looking at the ring status of the different components, it shows both instances with a roughly 50 - 50 ratio of ownership, so I'd assume it works as expected..?

[mimir.log](https://gist.github.com/SIMULATAN/62e36c5bf8bfcaa0c631408aa83992af) (includes the `mimir.yml` with substituted values at the top)
For some reason, the DNS query still fails, but it finds the other instances anyway...

**Reply by budiprihhastomo** [reply]
The same issue that I am experiencing, is there any fix or improvement regarding Mimir in the Docker Swarm environment?

**Reply by marvinnitz18** [reply]
i seem to have the exact same problem, but eth1 would be wrong for me, eth0 would be correct:
![image](https://github.com/grafana/mimir/assets/51240548/d18470d0-b36c-4b47-ba13-2603e901e371)

i will try to implement the workaround @SIMULATAN has provided now

**Reply by marvinnitz18** [reply]
for everyone using docker swarm, this i my solution i put the workaround into a wrapper.sh which i mounted into the container and defined it as the entrypoint in the stackfile:
`version: "3.7"
services:
  mimir1:
    image: grafana/mimir:2.11.0
    entrypoint: ["/bin/wrapper.sh"]`

wrapper.sh:
`#!/bin/sh
IP="$(ip addr show eth0 | grep inet | tr -s ' ' | cut -f 3 -d ' ' | cut -f 1 -d '/')"
exec /bin/mimir -config.file=/etc/mimir.yaml -memberlist.advertise-addr=$IP
`

**Reply by balzreber** [reply]
I have the exact same problem. I have a docker swarm and want to run 3 instances of mimir. So I somehow have to set the advertisement address dynamically. I tried the entripoint/wrapper hack listed here. Unfortunately, without success. I gladly share logs configs and so on, but one question first. Does the container that gets the wrapper mapped need to have /bin/sh installed itself? I assume so? As far as I can see, the current Mimir image (2.16) is bare bones and has no shell installed whatsoever. How do you get the wrapper to work? Any help is much appreciated. Thanks

**Reply by davbgallo** [reply]
@balzreber have you found a solution? I have the identical issue in docker swarm.

**Reply by balzreber** [reply]
@davbgallo Unfortunately I haven't. And I have moved on to a full on k8s cluster. So I didn't do any more research into this. I'm sorry.

**Reply by davbgallo** [reply]
I fixed my issue by switching to ubuntu/mimir and injecting the script :)

**Comment by ksingh-scogo** [comment]
@SIMULATAN @pracucci  encountering exact same issue with Mimir running on K8s 

```
ts=2023-12-24T12:18:35.380649131Z caller=log.go:245 level=debug msg="configured Transport is not a NodeAwareTransport and some features may not work as desired"
ts=2023-12-24T12:18:35.381232016Z caller=tcp_transport.go:405 level=debug component="memberlist TCPTransport" msg=FinalAdvertiseAddr advertiseAddr=10.244.2.176 advertisePort=7946
ts=2023-12-24T12:18:35.381724232Z caller=tcp_transport.go:405 level=debug component="memberlist TCPTransport" msg=FinalAdvertiseAddr advertiseAddr=10.244.2.176 advertisePort=7946

ts=2023-12-24T12:18:35.38402796Z caller=resolver.go:87 level=error msg="failed to lookup IP addresses" host=mimir-gossip-ring.grafana.svc.cluster.local. err="lookup mimir-gossip-ring.grafana.svc.cluster.local. on 10.0.0.10:53: no such host"

ts=2023-12-24T12:18:35.384052997Z caller=resolver.go:134 level=warn msg="IP address lookup yielded no results. No host found or no addresses found" host=mimir-gossip-ring.grafana.svc.cluster.local.

ts=2023-12-24T12:18:35.384066923Z caller=memberlist_client.go:541 level=info msg="memberlist fast-join starting" nodes_found=0 to_join=0
ts=2023-12-24T12:18:35.384078274Z caller=memberlist_client.go:561 level=warn msg="memberlist fast-join finished" joined_nodes=0 elapsed_time=2.201117ms

ts=2023-12-24T12:18:35.384090126Z caller=memberlist_client.go:573 level=info phase=startup msg="joining memberlist cluster" join_members=dns+mimir-gossip-ring.grafana.svc.cluster.local.:7946

ts=2023-12-24T12:18:35.385085489Z caller=resolver.go:87 level=error msg="failed to lookup IP addresses" host=mimir-gossip-ring.grafana.svc.cluster.local. err="lookup mimir-gossip-ring.grafana.svc.cluster.local. on 10.0.0.10:53: no such host"

ts=2023-12-24T12:18:35.385100887Z caller=resolver.go:134 level=warn msg="IP address lookup yielded no results. No host found or no addresses found" host=mimir-gossip-ring.grafana.svc.cluster.local.

ts=2023-12-24T12:18:35.3851131Z caller=memberlist_client.go:607 level=warn phase=startup msg="joining memberlist cluster" attempts=1 max_attempts=10 err="found no nodes to join"

```

<img width="374" alt="image" src="https://github.com/grafana/mimir/assets/76433381/f16692fb-2101-446c-8f18-4117d4cf7c0e">

**Reply by ksingh-scogo** [reply]
i am using mimir helm

**Reply by pracucci** [reply]
From the logs we can see:

> `err="lookup mimir-gossip-ring.grafana.svc.cluster.local. on 10.0.0.10:53: no such host"`

I suggest to run few sanity checks to begin with:

1. Does the `mimir-gossip-ring` K8S service exists in the `grafana` namespace? It should exist, as it's expected to be created as part of the Helm chart.
2. If you get a shell in one of the Mimir pods, can they resolve the service host `mimir-gossip-ring.grafana.svc.cluster.local`?
